[✨feat] StartDate(Filter) 구현

jsoonworld opened the pull request:
# 📄 Work Description  
- 근무 시작일을 의미하는 VO `StartDate`를 구현했습니다.  
- `StartDate`는 내부적으로 `Year`와 `Month` VO를 조합해 구성됩니다.  
- 각각의 VO는 생성 시 유효성 검사를 포함하고, JPA `@Embeddable`로 설계되어 도메인에 안전하게 통합됩니다.  
- VO 간 중첩 매핑을 위해 `@Embedded`, `@AttributeOverride`를 활용해 컬럼명을 명시적으로 지정했습니다.

# 💭 Thoughts  
- 단순히 `Int`로 처리하던 연도와 월 값을 각각 `Year`, `Month` VO로 분리함으로써 도메인 내 의미를 명확히 표현할 수 있었습니다.  
- 각 VO는 스스로 유효성을 검증하고, 값 객체로서 동일성 비교가 가능하도록 구현되어 책임이 명확히 분리되었습니다.  
- 특히 `StartDate`라는 VO로 두 값을 묶음으로써, 단순 값 조합 이상의 의미(근무 시작일)를 코드 레벨에서 드러낼 수 있게 되었고, 이후 `GraduationDate`, `EndDate` 등 다양한 날짜 표현에 일관된 구조로 확장할 수 있는 기반이 되었습니다.  
- `Year`와 `Month`를 개별 VO로 만든 이유는 재사용 가능성과 검증 책임 분리 때문이며, 이를 조합해 `StartDate`라는 유의미한 도메인 객체를 구성하는 방향이 도메인 응집도 측면에서 적절하다고 판단했습니다.

# ✅ Testing Result  
![스크린샷 2025-05-18 오전 1 06 33](https://github.com/user-attachments/assets/a398acbf-02f2-4bff-a4b9-214be4d975e4)


# 🗂 Related Issue  
- closes #32

jsoonworld committed:
feat(Month) : 월 VO 구현

1~12 범위의 월 값을 표현하는 Month VO를 정의했습니다.
유효성 검증을 포함하고, equals 및 hashCode를 오버라이드하여 값 객체로 사용할 수 있도록 구성했습니다.

jsoonworld committed:
feat(Year) : 연도 VO 구현

1900보다 큰 값을 표현하는 Year VO를 정의했습니다.
유효성 검증을 포함하고, equals 및 hashCode를 오버라이드하여 값 객체로 사용할 수 있도록 구성했습니다.

jsoonworld committed:
feat(StartDate) : 근무 시작일 VO 구현

Year와 Month를 조합하여 근무 시작일을 표현하는 StartDate VO를 구현했습니다.
JPA 임베디드 매핑을 위해 AttributeOverride를 적용하고, equals 및 hashCode를 오버라이드하여 값 객체로 사용할 수 있도록 구성했습니다.

jsoonworld committed:
test(Month) : 월 VO 생성 및 유효성 검증 테스트 추가

1부터 12 사이의 값을 정상적으로 생성하는 경우와
유효하지 않은 월(0 이하, 13 이상)에 대해 예외가 발생하는 경우를 검증하는 테스트를 작성했습니다.

jsoonworld committed:
test(StartDate) : StartDate VO 생성 테스트 추가

Year와 Month를 기반으로 StartDate 객체를 생성하고,
내부 값이 정확히 설정되는지를 검증하는 테스트를 작성했습니다.

jsoonworld committed:
test(Year) : Year VO 생성 및 유효성 검증 테스트 추가

유효한 연도로 Year 객체를 생성할 수 있는지와
1900년 이하 입력 시 예외가 발생하는 경우를 검증하는 테스트를 작성했습니다.

jsoonworld commented on src/main/kotlin/com/terning/server/kotlin/domain/filter/StartDate.kt:
```diff
@@ -0,0 +1,35 @@
+    val month: Month,
+) {
+    companion object {
+        fun of(
```
StartDate는 단순 값을 감싸는 VO(Year, Month)와 달리 두 개의 파라미터를 조합해 의미 있는 도메인 객체를 생성하기 때문에, 생성 의미를 더 명확히 드러내기 위해 of를 선택했습니다.

from은 일반적으로 단일 원시값을 감싸는 VO에 주로 사용하고 있고, of는 여러 값을 조합하는 팩토리 메서드에 일관되게 적용하고자 했습니다. 따라서 API 레벨에서는 의도적으로 구분하고 있습니다 🙂

jsoonworld committed:
refactor(StartDate) : VO 내 컬럼명을 제거하고 매핑 책임을 StartDate로 위임

- Year, Month VO 내부의 @Column name 속성을 제거하여 재사용성을 높였습니다.
- DB 컬럼명은 StartDate에서 @AttributeOverride를 통해 명시적으로 지정했습니다.
- equals 메서드를 간결한 표현식으로 개선했습니다.

jsoonworld committed:
chore(StartDate): ktlint 포맷 적용 및 중복 equals 메서드 제거

jsoonworld commented on src/main/kotlin/com/terning/server/kotlin/domain/filter/Year.kt:
```diff
@@ -0,0 +1,32 @@
+
+@Embeddable
+class Year private constructor(
+    @Column(name = "start_year", nullable = false)
```
피드백 감사합니다!
말씀해주신 대로 VO 내부에서 컬럼명을 제거하고, 엔티티 쪽에서 @AttributeOverride로 명시하도록 수정하겠습니다.
이렇게 하면 재사용성도 높아지고 설계도 더 깔끔해질 것 같아요!

jsoonworld commented on src/main/kotlin/com/terning/server/kotlin/domain/filter/Year.kt:
```diff
@@ -0,0 +1,32 @@
+
+import jakarta.persistence.Column
+import jakarta.persistence.Embeddable
+
```
피드백 감사합니다!
코드 자체 네이밍이 명확하면 주석이나 별도의 설명은 꼭 필요하지 않을 것 같아요.
현재로서는 도메인 의도가 잘 드러난다고 생각해서,
필요할 때 그때그때 고민해보는 방향으로 가면 좋겠습니다 :)

jsoonworld commented on src/main/kotlin/com/terning/server/kotlin/domain/filter/Month.kt:
```diff
@@ -0,0 +1,32 @@
+
+@Embeddable
+class Month private constructor(
+    @Column(name = "start_month", nullable = false)
```
피드백 감사합니다!
Year와 Month VO 내부에 있던 구체적인 컬럼명 지정(@Column(name = "..."))을 제거하고,
이제 해당 VO를 포함하는 엔티티나 VO(StartDate) 쪽에서 @AttributeOverride로 명확하게 컬럼명을 지정하도록 수정했습니다.

leeeyubin commented on src/main/kotlin/com/terning/server/kotlin/domain/filter/Month.kt:
```diff
@@ -0,0 +1,32 @@
+    @Column(nullable = false)
+    val value: Int,
+) {
+    init {
+        require(value in 1..12) { INVALID_MONTH_MESSAGE.format(value) }
+    }
```
1과 12도 상수화를 통해 어떤 걸 의미하는 지 나타내주면 좋을 것 같아요!

jsoonworld replied:
좋은 피드백 감사합니다! 🙌  
말씀해주신 것처럼 `1`과 `12`를 상수로 추출하면서  
각 값이 의미하는 바를 좀 더 명확하게 드러낼 수 있도록 개선해보았습니다.

특히 조건식이 `MIN_MONTH..MAX_MONTH` 형태로 바뀌니  
코드 읽을 때도 훨씬 자연스럽게 의미가 전달되는 것 같아요!  
꼼꼼하게 봐주셔서 감사합니다 😊

leeeyubin commented on src/main/kotlin/com/terning/server/kotlin/domain/filter/StartDate.kt:
```diff
@@ -0,0 +1,29 @@
+
+@Embeddable
+class StartDate private constructor(
+    @Embedded
+    @AttributeOverride(name = "value", column = Column(name = "start_year", nullable = false))
+    val year: Year,
+    @Embedded
+    @AttributeOverride(name = "value", column = Column(name = "start_month", nullable = false))
+    val month: Month,
```
VO에 값이 두 개 이상이 들어가는 경우 이렇게 작성하는군요!
내부적으로 value라는 필드로 정의하고 각각 `start_year`와 `start_month`로 매핑해주고 있네요! 배워갑니다!!

jsoonworld replied:
맞아요! VO를 엔티티에 `@Embedded`로 사용할 때,  
내부 필드명이 동일하거나 겹칠 수 있어서 `@AttributeOverride`로 매핑 컬럼명을 명시해주는 방식이에요 🙌

예를 들어 `Year`, `Month` 모두 내부적으로 `value`라는 프로퍼티를 갖고 있다 보니,  
그대로 두면 JPA가 어떤 필드가 어떤 컬럼에 대응되는지 알 수 없어서 에러가 나요.

그래서 `@AttributeOverride(name = "value", column = Column(name = "..."))` 형식으로  
각 VO의 `value` 필드를 `start_year`, `start_month` 컬럼에 각각 명확히 매핑해주고 있어요!

VO를 설계할 때 이렇게 **불변성과 의미를 가진 객체**로 만들면서도,  
JPA 매핑에서는 `@Embedded` + `@AttributeOverride` 조합으로 잘 사용할 수 있어서 자주 쓰이는 방식이에요 😊  
이 부분 눈여겨봐주셔서 감사드려요! 함께 나눌 수 있어 더 좋았습니다 🙏

leeeyubin commented on src/test/kotlin/com/terning/server/kotlin/domain/filter/MonthTest.kt:
```diff
@@ -0,0 +1,26 @@
+import org.junit.jupiter.api.assertThrows
+
+class MonthTest {
+    @Test
+    @DisplayName("1부터 12 사이 값으로 Month를 생성할 수 있다")
+    fun createValidMonth() {
+        val month = Month.from(6)
+        assertThat(month.value).isEqualTo(6)
+    }
```
테스트 코드를 보다가 든 생각인데 given, when, then을 주석으로 분류해서 나타내면 직관적으로 보기에 좋을 것 같기도 하네요!

jsoonworld replied:
좋은 제안 감사합니다! 🙌  
이 테스트는 워낙 간단해서 그동안은 given-when-then 주석 없이 작성했었는데,  
말씀해주신 것처럼 테스트 전반에 걸쳐 일관된 포맷으로 작성하면  
읽는 사람 입장에서도 흐름이 더 명확하게 드러날 것 같아요!

특히 테스트 케이스가 많아질수록 `given`, `when`, `then` 주석이 있는 것과 없는 것의 가독성 차이가 커지더라구요 😅  
이 부분은 반영해서 개선해보겠습니다! 좋은 인사이트 공유해주셔서 감사해요 😊

leeeyubin commented on src/test/kotlin/com/terning/server/kotlin/domain/filter/MonthTest.kt:
```diff
@@ -0,0 +1,26 @@
+
+    @Test
+    @DisplayName("0 이하 또는 13 이상의 월은 예외가 발생한다")
+    fun throwExceptionWhenMonthIsInvalid() {
+        assertThrows<IllegalArgumentException> {
+            Month.from(0)
+        }
+        assertThrows<IllegalArgumentException> {
+            Month.from(13)
+        }
```
예외가 발생하는 테스트 코드는 `assertThrows`를 사용하는군요! 👍

jsoonworld replied:
맞아요! 예외를 검증할 때는 `assertThrows`나 `assertThatThrownBy`를 상황에 맞게 사용하고 있어요 🙌  
이 테스트에서는 예외 타입만 확인하면 충분해서 JUnit의 `assertThrows`를 간단히 썼고,  
반면에 이전 테스트들(ex. `ColorTest`, `WorkingPeriodTest`)에서는  
예외의 **에러 메시지나 에러 코드까지 함께 검증**해야 했기 때문에  
AssertJ의 `assertThatThrownBy`와 `isInstanceOfSatisfying`을 함께 사용했었어요.

각 테스트 목적에 맞게 도구를 선택했지만,  
테스트 흐름을 더 명확하게 보여주기 위해 `expect` 주석 포맷을 같이 맞춰보는 것도 좋을 것 같네요!  
피드백 감사드려요 😊

leeeyubin replied:
오호 친절한 설명 감사합니다🚀🚀
말씀해주신 내용들도 공부해볼게요!!

leeeyubin reviewed (approved):
VO를 통해 구현하신 점 잘 보았습니다!! 참고해갑니당~~

jsoonworld committed:
refactor(Month) : 월 유효성 범위를 상수로 추출하여 가독성 및 유지보수성 개선

- 1~12의 매직 넘버를 MIN_MONTH, MAX_MONTH 상수로 추출
- 에러 메시지도 상수를 활용해 의미 전달력 향상
- 도메인 유효성 검증의 의도를 더 명확히 표현

jsoonworld committed:
refactor(test) : 도메인 테스트에 given-when-then 포맷 및 명확한 표현 적용

- 모든 도메인 테스트에 given/when/then 또는 expect 주석 포맷을 적용하여 테스트 흐름 가독성 향상
- label → color 등 변경된 도메인 필드에 맞춰 DisplayName 및 테스트 설명 수정
- 예외 검증 방식 개선 및 테스트 명세 명확화
- 테스트 구조의 일관성을 위해 필요한 경우 @Nested 및 파라미터 설명 포맷 정리

jsoonworld committed:
merge(feat/#30 → feat/#32) : 이전 브랜치 변경사항 반영을 위한 선병합

jsoonworld committed:
chore(format) : ktlintFormat 적용으로 코드 스타일 정렬
